examples/jsm/renderers/webgl/nodes/WebGLNodes.js

```diff
@@ -0,0 +1,45 @@
+import { WebGLNodeBuilder } from './WebGLNodeBuilder.js';
+
+import { Material } from '../../../../../build/three.module.js';
+
+function addCodeAfterSnippet( source, snippet, code ) {
+	
+	const index = source.indexOf( snippet );
+	
+	if ( index !== -1 ) {
+		
+		const start = source.substring( 0, index + snippet.length );
+		const end = source.substring( index + snippet.length );
+		
+		return `${start}\n${code}\n${end}`;
+		
+	}
+	
+	return source;
+	
+}
+
+Material.prototype.onBeforeCompile = function( parameters, renderer ) {
```

Comment:
This seems a bit problematic. You are monkey-patching `onBeforeCompile()` here which means this code will affect all materials when `WebGLNodes` is included (even if a material does not use the node system). And when users implement `onBeforeCompile()` on app level, they will overwrite this code. 

I would like to get away from `onBeforeCompile()` in the node context and introduce a new (private) callback intended for the node system which enables the above code injection. In this way, it does not conflict with `onBeforeCompile()`. I would also suggest that only materials with an enabled flag ( e.g. `Material.useNodes`) are affected by this new callback.

Besides, the new node system also has to provide something like `Node.getHash()` again so `customProgramCacheKey()` can identify the correct program.

@mrdoob `onBeforeCompile()` was introduced for the current material system. I think such a callback is counterproductive in a node system where custom materials should be coded/configured differently. What are your plans with `onBeforeCompile()` in context of node materials? Do you already have an idea/vision how users supposed to modify built-in shader code?

Reply:
Using that already exists it is possible to do so as well:
```js
class MeshStandardNodeMaterial extends MeshStandardMaterial {
	
	constructor( parameters ) {
		
		super( parameters );
		
		this.colorNode = null;

		this.onBeforeCompile = ...
			
	}
	
}

```

Reply:
Yes, extending build-in materials and creating node pendants would be an alternative way for having a flag `Material.useNodes`. However, `onBeforeCompile()` would still be used by the node system and hence reserved. This is not necessarily in issue but it could be for the reason mentioned in https://github.com/mrdoob/three.js/pull/21117#discussion_r571648032.

Let's see what @mrdoob thinks about this topic. These are quite fundamental design decisions for the engine and would be good to think through the possible options.

Reply:
Yes, I think it's better to not rely on `onBeforeCompile()` whenever possible.

Reply:
This other approach can be an alternative too:
If we bring some `nodes` to the `core`, we won't need to create new callback functions.

https://github.com/mrdoob/three.js/blob/d37f10fe861823d76d4f2b0b6887579b12d29815/src/renderers/WebGLRenderer.js#L1367-L1371

Reply:
I vote for a callback function for now since it's less hardwired. Maybe simply:
```js
material.onBuild( parameters, renderer );
```

Reply:
> material.onBuild( parameters, renderer );

It seems like two functions competing with each other?

Reply:
My intention is that `onBeforeCompile()` is no longer used by node material. Since the material system requires some sort of trigger that executes the build, I thought `onBuild()` (or maybe just `build()`) is sufficient.

Reply:
If follow replace `Material.prototype.onBuild = ...` I vote for `onBuild`.

If follow a material extended  like ( https://github.com/mrdoob/three.js/pull/21117#discussion_r571684436 ) I still don't see the need for the moment. I think that anyone who would use `onBeforeCompile` to customize a shader would not use `NodeMaterial` and vice versa...

Reply:
Will the parameters passed to `onBuild` be the same as the parameters passed to `onBeforeCompile`?

Reply:
Right now yes although we could change/enhance it since it would be something new.

Reply:
Right now yes although we could change/enhance it since it would be something new.

Reply:
To sum up: The PR tries to enhance build-in materials with node logic. Right now, the PR globally overwrites `Material.onBeforeCompile()` in `WebGLNodes.js` which means it affects all materials no matter if they use node logic or not. IMO, this approach is too invasive.

There are two alternatives now to implement this more gracefully (@sunag please correct/enhance this section if I'm wrong here^^):

- Create for all built-in material node material classes like `MeshStandardNodeMaterial` and stick with `onBeforeCompile()`.
- Introduce `Material.onBuild()` and let the node material overwrite this method in `WebGLNodes.js`. `Material.onBuild()` would be considered as no public part of the API.

Reply:
@Mugen87  Exactly. 
Just an extra consideration: `NodeMaterial` of `WebGPU` not use extended materials like `MeshStandardNodeMaterial` but extended properties in core materials like second option `Material.onBuild()` should be .

Reply:
@mrdoob Do you think you can choose one of the options mentioned in https://github.com/mrdoob/three.js/pull/21117/files#r580199457 😇 ? It would be good if we could merge this PR in this dev cycle so the new node material system works with WebGPU and WebGL.

Reply:
@sunag Do you mind implementing this PR with `Material.onBuild()`. I can't speak for @mrdoob but I _feel_ this approach will be more acceptable.

Reply:
I also belive that `Material.onBuild()` makes more sense with the evolution of `WebGPU`

Reply:
Ops, missed this one. Yes `Material.onBuild()` sounds good to me.

Reply:
@sunag Do you mind updating the PR and introduce `Material.onBuild()`?